src/Disks/ObjectStorages/Cached/CachedObjectStorage.cpp

```diff
@@ -63,21 +62,60 @@ bool CachedObjectStorage::exists(const StoredObject & object) const
     return object_storage->exists(object);
 }
 
-std::unique_ptr<ReadBufferFromFileBase> CachedObjectStorage::readObjects( /// NOLINT
-    const StoredObjects & objects,
-    const ReadSettings & read_settings,
-    std::optional<size_t> read_hint,
-    std::optional<size_t> file_size) const
-{
-    return object_storage->readObjects(objects, patchSettings(read_settings), read_hint, file_size);
-}
-
 std::unique_ptr<ReadBufferFromFileBase> CachedObjectStorage::readObject( /// NOLINT
     const StoredObject & object,
     const ReadSettings & read_settings,
     std::optional<size_t> read_hint,
     std::optional<size_t> file_size) const
 {
+    if (read_settings.enable_filesystem_cache)
```

Comment:
The purpose of this PR is to get rid of this mess that we added cached buffer in a very implicit way - in ReadBufferFromRemoteFSGather where we understood that we need to add cached buffer via ReadSettings. And in general it makes the buffers chain creation much more clear.
This refactoring will also allow:
(1) to merge https://github.com/ClickHouse/ClickHouse/pull/52021
(2) to get rid of ugly inconsistency in code in https://github.com/ClickHouse/ClickHouse/pull/70135